pyro/contrib/forecast/guides.py

```diff
@@ -0,0 +1,139 @@
+
+        self._event_dims = {}
+        self._is_temporal = {}
+        self.locs = PyroModule()
+        self.scales = PyroModule()
+        self.f_locs = PyroModule()
+        self.f_scales = PyroModule()
+
+        # Initialize guide params.
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            with torch.no_grad():
+                loc = biject_to(site["fn"].support).inv(site["value"].detach())
+                scale = torch.full_like(loc, self._init_scale)
+            event_dim = site["fn"].event_dim + loc.dim() - site["value"].dim()
+            self._event_dims[name] = event_dim
+            self._is_temporal[name] = any(f.name == "time" for f in site["cond_indep_stack"])
+            if self._is_temporal[name]:
+                # Split parameters into time and frequency part.
+                loc *= 0.5
+                scale *= 0.5 ** 0.5
+                f_loc = dct(loc, dim=-1 - event_dim)
+                f_scale = scale.clone()
+
+                _deep_setattr(self.f_locs, name, PyroParam(f_loc, constraints.real, event_dim))
+                _deep_setattr(self.f_scales, name, PyroParam(f_scale, constraints.positive, event_dim))
+            _deep_setattr(self.locs, name, PyroParam(loc, constraints.real, event_dim))
+            _deep_setattr(self.scales, name, PyroParam(scale, constraints.positive, event_dim))
+
+    def forward(self, *args, **kwargs):
+        if self.prototype_trace is None:
+            self._setup_prototype(*args, **kwargs)
+
+        plates = self._create_plates()
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            event_dim = self._unconstrained_event_dims[name]
+            transform = biject_to(site["fn"].support)
+
+            with ExitStack() as stack:
+                for frame in site["cond_indep_stack"]:
+                    if frame.vectorized and frame.name != "time":
+                        stack.enter_context(plates[frame.name])
+
+                loc = _deep_getattr(self.locs, name)
+                scale = _deep_getattr(self.scales, name)
+
+                # For temporal sites, add uncertainty in the frequency domain.
+                if self._is_temporal[name]:
+                    with plates["frequency"]:
+                        f_loc = _deep_getattr(self.f_locs, name)
+                        f_scale = _deep_getattr(self.f_scales, name)
+                        f = pyro.sample(name + "_frequency",
+                                        dist.Normal(f_loc, f_scale).to_event(event_dim),
+                                        infer={"is_auxiliary": True})
+                    loc = loc + idct(f, dim=-1 - event_dim)
+                    stack.enter_context(plates["time"])
+
+                # Sample from a transformed diagonal normal.
+                base_dist = dist.Normal(loc, scale).to_event(event_dim)
+                pyro.sample(name, dist.TransformedDistribution(base_dist, transform))
```

Comment:
@fehiepsi @martinjankowiak can you help me convince myself that this auxiliary variable method is correct?